ext/compressors/zstd/zstd_compress.c

```diff
@@ -0,0 +1,334 @@
+typedef struct {
+	WT_COMPRESSOR compressor;		/* Must come first */
+
+	WT_EXTENSION_API *wt_api;		/* Extension API */
+
+	int	 compression_level;		/* compression level */
+} ZSTD_COMPRESSOR;
+
+#ifdef WORDS_BIGENDIAN
+/*
+ * zstd_bswap64 --
+ *	64-bit unsigned little-endian to/from big-endian value.
+ */
+static inline uint64_t
+zstd_bswap64(uint64_t v)
+{
+	return (
+	    ((v << 56) & 0xff00000000000000UL) |
+	    ((v << 40) & 0x00ff000000000000UL) |
+	    ((v << 24) & 0x0000ff0000000000UL) |
+	    ((v <<  8) & 0x000000ff00000000UL) |
+	    ((v >>  8) & 0x00000000ff000000UL) |
+	    ((v >> 24) & 0x0000000000ff0000UL) |
+	    ((v >> 40) & 0x000000000000ff00UL) |
+	    ((v >> 56) & 0x00000000000000ffUL)
+	);
+}
+#endif
+
+/*
+ * zstd_error --
+ *	Output an error message, and return a standard error code.
+ */
+static int
+zstd_error(WT_COMPRESSOR *compressor,
+    WT_SESSION *session, const char *call, size_t error)
+{
+	WT_EXTENSION_API *wt_api;
+
+	wt_api = ((ZSTD_COMPRESSOR *)compressor)->wt_api;
+
+	(void)wt_api->err_printf(wt_api, session,
+	    "zstd error: %s: %s", call, ZSTD_getErrorName(error));
+	return (WT_ERROR);
+}
+
+/*
+ *  zstd_compress --
+ *	WiredTiger Zstd compression.
+ */
+static int
+zstd_compress(WT_COMPRESSOR *compressor, WT_SESSION *session,
+    uint8_t *src, size_t src_len,
+    uint8_t *dst, size_t dst_len,
+    size_t *result_lenp, int *compression_failed)
+{
+	ZSTD_COMPRESSOR *zcompressor;
+	size_t zstd_ret;
+
+	zcompressor = (ZSTD_COMPRESSOR *)compressor;
+
+	/* Compress, starting past the prefix bytes. */
+	zstd_ret = ZSTD_compress(
+	    dst + sizeof(size_t), dst_len - sizeof(size_t),
+	    src, src_len, zcompressor->compression_level);
+
+	/*
+	 * If compression succeeded and the compressed length is smaller than
+	 * the original size, return success.
+	 */
+	if (!ZSTD_isError(zstd_ret) && zstd_ret + sizeof(size_t) < src_len) {
+		*result_lenp = zstd_ret + sizeof(size_t);
+		*compression_failed = 0;
+
+		/*
+		 * On decompression, Zstd requires an exact compressed byte
+		 * count (the current value of zstd_ret). WiredTiger does not
+		 * preserve that value, so save zstd_ret at the beginning of
+		 * the destination buffer.
+		 *
+		 * Store the value in little-endian format.
+		 */
+#ifdef WORDS_BIGENDIAN
+		zstd_ret = zstd_bswap64(zstd_ret);
+#endif
+		*(size_t *)dst = zstd_ret;
+		return (0);
+	}
+
+	*compression_failed = 1;
+	return (ZSTD_isError(zstd_ret) ?
+	    zstd_error(compressor, session, "ZSTD_compress", zstd_ret) : 0);
+}
+
+/*
+ * zstd_decompress --
+ *	WiredTiger Zstd decompression.
+ */
+static int
+zstd_decompress(WT_COMPRESSOR *compressor, WT_SESSION *session,
+    uint8_t *src, size_t src_len,
+    uint8_t *dst, size_t dst_len,
+    size_t *result_lenp)
+{
+	size_t zstd_ret;
+
+	/*
+	 * Retrieve the saved length, handling little- to big-endian conversion
+	 * as necessary.
+	 */
+	src_len = *(size_t *)src;
+#ifdef WORDS_BIGENDIAN
+	src_len = zstd_bswap64(src_len);
+#endif
```

Comment:
> Likewise here, I think we need to sanity check the bits from disk in case they are corrupted.

Included in commit da25f35.